File: backend/app/schemas/eda.py
"""
backend/app/schemas/eda.py
Pydantic schemas for automated Exploratory Data Analysis (EDA) reports.
"""
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field


class CorrelationPair(BaseModel):
    col1: str
    col2: str
    correlation: float
    strength: str  # "Strong Positive", "Moderate Positive", "Weak", "Moderate Negative", "Strong Negative"


class CorrelationMatrix(BaseModel):
    columns: List[str]
    matrix: List[List[float]]   # N×N grid, row-major
    top_pairs: List[CorrelationPair]


class HistogramBin(BaseModel):
    bin_start: float
    bin_end: float
    count: int


class DistributionProfile(BaseModel):
    column: str
    mean: float
    median: float
    std: float
    skewness: float
    kurtosis: float
    min: float
    max: float
    bins: List[HistogramBin]


class CategoricalFrequency(BaseModel):
    value: str
    count: int
    percentage: float


class CategoricalBreakdown(BaseModel):
    column: str
    total_distinct: int
    frequencies: List[CategoricalFrequency]   # up to top-15 + "Other" rollup
    other_count: int
    other_percentage: float


class ScatterPoint(BaseModel):
    x: float
    y: float


class PairwiseScatter(BaseModel):
    x_column: str
    y_column: str
    correlation: float
    strength: str
    points: List[ScatterPoint]   # capped at 300 points for performance


class EDANarrativeSummary(BaseModel):
    overview: str
    key_findings: List[str]


class EDAReportResponse(BaseModel):
    dataset_id: str
    dataset_name: str
    row_count: int
    column_count: int
    numeric_column_count: int
    categorical_column_count: int
    correlations: Optional[CorrelationMatrix] = None
    distributions: List[DistributionProfile]
    categorical_breakdowns: List[CategoricalBreakdown]
    pairwise_scatters: List[PairwiseScatter]
    narrative_summary: EDANarrativeSummary
    computed_at: Optional[str] = None
    is_cached: bool = False

File: frontend/types/eda.ts
export interface CorrelationPair {
  col1: string;
  col2: string;
  correlation: number;
  strength: string;
}

export interface CorrelationMatrix {
  columns: string[];
  matrix: number[][];
  top_pairs: CorrelationPair[];
}

export interface HistogramBin {
  bin_start: number;
  bin_end: number;
  count: number;
}

export interface DistributionProfile {
  column: string;
  mean: number;
  median: number;
  std: number;
  skewness: number;
  kurtosis: number;
  min: number;
  max: number;
  bins: HistogramBin[];
}

export interface CategoricalFrequency {
  value: string;
  count: number;
  percentage: number;
}

export interface CategoricalBreakdown {
  column: string;
  total_distinct: number;
  frequencies: CategoricalFrequency[];
  other_count: number;
  other_percentage: number;
}

export interface ScatterPoint {
  x: number;
  y: number;
}

export interface PairwiseScatter {
  x_column: string;
  y_column: string;
  correlation: number;
  strength: string;
  points: ScatterPoint[];
}

export interface EDANarrativeSummary {
  overview: string;
  key_findings: string[];
}

export interface EDAReportResponse {
  dataset_id: string;
  dataset_name: string;
  row_count: number;
  column_count: number;
  numeric_column_count: number;
  categorical_column_count: number;
  correlations: CorrelationMatrix | null;
  distributions: DistributionProfile[];
  categorical_breakdowns: CategoricalBreakdown[];
  pairwise_scatters: PairwiseScatter[];
  narrative_summary: EDANarrativeSummary;
  computed_at: string | null;
  is_cached: boolean;
}
